src/mbgl/map/transform.cpp

```diff
@@ -519,130 +617,146 @@ ProjectionMode Transform::getProjectionMode() const {
 
 // MARK: - Transition
 
-void Transform::startTransition(const CameraOptions& camera,
-                                const AnimationOptions& animation,
-                                const std::function<void(double)>& frame,
-                                const Duration& duration) {
-    if (transitionFinishFn) {
-        transitionFinishFn();
+void Transform::animationFinishFrame(Animation& animation) {
+    if (animation.finished) {
+        return;
     }
 
+    if (animation.options.transitionFinishFn) {
+        animation.options.transitionFinishFn();
+    }
+
+    animation.finished = true;
+
+    if (animation.anchor) animation.anchor = std::nullopt;
+
+    observer.onCameraDidChange(animation.isAnimated() ? MapObserver::CameraChangeMode::Animated
+                                                      : MapObserver::CameraChangeMode::Immediate);
+}
+
+void Transform::startTransition(const CameraOptions& camera, const Duration& duration, Animation& animation) {
     bool isAnimated = duration != Duration::zero();
     observer.onCameraWillChange(isAnimated ? MapObserver::CameraChangeMode::Animated
                                            : MapObserver::CameraChangeMode::Immediate);
 
     // Associate the anchor, if given, with a coordinate.
     // Anchor and center points are mutually exclusive, with preference for the
     // center point when both are set.
-    std::optional<ScreenCoordinate> anchor = camera.center ? std::nullopt : camera.anchor;
-    LatLng anchorLatLng;
-    if (anchor) {
-        anchor->y = state.getSize().height - anchor->y;
-        anchorLatLng = state.screenCoordinateToLatLng(*anchor);
+    if (!camera.center && camera.anchor) {
+        animation.anchor = camera.anchor;
+        animation.anchor->y = state.getSize().height - animation.anchor->y;
+        animation.anchorLatLng = state.screenCoordinateToLatLng(*animation.anchor);
     }
 
-    transitionStart = Clock::now();
-    transitionDuration = duration;
+    if (!isAnimated) {
+        activeAnimation = false;
+        updateTransitions(Clock::now());
+    }
+}
 
-    transitionFrameFn = [isAnimated, animation, frame, anchor, anchorLatLng, this](const TimePoint now) {
-        float t = isAnimated ? (std::chrono::duration<float>(now - transitionStart) / transitionDuration) : 1.0f;
-        if (t >= 1.0) {
-            frame(1.0);
-        } else {
-            util::UnitBezier ease = animation.easing ? *animation.easing : util::DEFAULT_TRANSITION_EASE;
-            frame(ease.solve(t, 0.001));
-        }
+bool Transform::inTransition() const {
+    return properties.latlng.set || properties.zoom.set || properties.bearing.set || properties.padding.set ||
+           properties.pitch.set;
+}
 
-        if (anchor) state.moveLatLng(anchorLatLng, *anchor);
+void Transform::updateTransitions(const TimePoint& now) {
+    if (!activeAnimation) {
+        activeAnimation = true;
+
+        bool panning = false, scaling = false, rotating = false;
```

Comment:
Nit: we tend not to use multiple variable declarations.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-name-one